### ISSUE TITLE
no-shadow: allow array methods to reference and manipulate variables declared outside

### ISSUE THREAD

**Issue by yogurtandjam** [eslint/eslint#12914]
**What rule do you want to change?**
no-shadow

**Does this change cause the rule to produce more or fewer warnings?**
fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
new default behavior

**Please provide some example code that this change will affect:**
```js
const filterSameItems = notifications => {
  const map = {};
  return notifications.filter(notification => {
    if (map[notification.message]) {
      return false;
    }
    map[notification.message] = true;
    return true;
  });
};
```

**What does the rule currently do for this code?**
Returns a no-shadow error "map is already declared in the upper scope" thinking that we are attempting to re-assign or implicitly declare map.

**What will the rule do after it's changed?**
Allow callbacks to reference and manipulate variables declared outside of their scope

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by mdjermanovic** [comment]
Hi @yogurtandjam, thanks for the issue!

Can you please clarify the proposal, I'm trying to reproduce a no-shadow error with the example you provided, but the code is lint-free:

```js
/* eslint no-shadow: error */

const filterSameItems = notifications => {
  const map = {};
  return notifications.filter(notification => {
    if (map[notification.message]) {
      return false;
    }
    map[notification.message] = true;
    return true;
  });
};
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXNoYWRvdzogZXJyb3IgKi9cblxuY29uc3QgZmlsdGVyU2FtZUl0ZW1zID0gbm90aWZpY2F0aW9ucyA9PiB7XG4gIGNvbnN0IG1hcCA9IHt9O1xuICByZXR1cm4gbm90aWZpY2F0aW9ucy5maWx0ZXIobm90aWZpY2F0aW9uID0+IHtcbiAgICBpZiAobWFwW25vdGlmaWNhdGlvbi5tZXNzYWdlXSkge1xuICAgICAgcmV0dXJuIGZhbHNlO1xuICAgIH1cbiAgICBtYXBbbm90aWZpY2F0aW9uLm1lc3NhZ2VdID0gdHJ1ZTtcbiAgICByZXR1cm4gdHJ1ZTtcbiAgfSk7XG59OyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

**Comment by yogurtandjam** [comment]
@mdjermanovic you're right - sorry this is why I probably shouldn't submit issue reports at midnight. Thanks for taking your time on this and once again sorry!
